src/fair/allocation.py

```diff
@@ -186,39 +188,57 @@ def build_exchange_graph(X, items, agents):
                     exchange_graph.add_edge(item_index, item_2_index)
     return exchange_graph
 
+def get_multiple_agents_desired_items(agents_indexes,agents,items):
+    lis = []
+    for agent_index in agents_indexes:
+        agent = agents[agent_index]
+        lis = lis + agent.get_desired_items_indexes(items)
+    return list(set(lis))
+
+def get_multiple_agents_bundles(agents_indexes,X):
+    lis = []
+    for agent_index in agents_indexes:
+        lis = lis + get_bundle_indexes_from_allocation_matrix(X,agent_index)
+    return list(set(lis))
+
+
 
 def update_exchange_graph(X, G, path_og, agents, items, agents_involved):
     path = path_og.copy()
     path = path[1:-1]
     last_item = path[-1]
     if X[last_item, len(agents)] == 0:
         G.remove_edge(last_item, "t")
-    for agent_index in agents_involved:
```

Comment:
I understand your desire to introduce the methods above to break up the code block below. But perhaps we can talk some time about how to more comprehensively decompose this. One idea would be to extract all the logic inside the inner loop into one or more separate method. Having separated that logic from the loops, it would hopefully be more easily testable.